runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkBatchPortablePipelineTranslator.java

```diff
@@ -111,6 +112,25 @@ public void translate(final RunnerApi.Pipeline pipeline, SparkTranslationContext
     QueryablePipeline p =
         QueryablePipeline.forTransforms(
             pipeline.getRootTransformIdsList(), pipeline.getComponents());
+    for (PipelineNode.PTransformNode transformNode : p.getTopologicallyOrderedTransforms()) {
+      // Pre-scan pipeline to count which pCollections are consumed as inputs more than once so
+      // their corresponding RDDs can later be cached.
+      for (String inputId : transformNode.getTransform().getInputsMap().values()) {
+        context.incrementConsumptionCountBy(inputId, 1);
+      }
+      // Executable stage consists of two parts: computation and extraction. This means the result
+      // of computation is an intermediate RDD, which we might also need to cache.
```

Comment:
Is this preferable to extracting the RDD corresponding to the (potentially expensive) extraction? (In that case, rather than special casing ExecutableStage, we could simply increment all outputs of multi-output transforms.)

Reply:
We have to cache the output of both the executable stage itself and the extraction function in order to fully prevent re-computation (tested in `testExecStageWithMultipleOutputs` and `testExecStageWithMultipleConsumers`, respectively). The immediate output of the stage (ie the input to extraction) doesn't correspond to any RunnerApi-level pCollection, being a runner-specific detail, which is why it's special-cased here.